File: src/templates/skills/bbva-brand/scripts/ts/generate_pptx.ts
#!/usr/bin/env npx tsx
/**
 * generate_pptx.ts — BBVA-branded PPTX generator (pptxgenjs, DEFAULT runtime).
 *
 * Usage:
 *   npx tsx generate_pptx.ts --content content.json --output output.pptx
 *   npx tsx generate_pptx.ts --title "My Deck" --output output.pptx
 *
 * content.json schema: { title, subtitle?, author?, sections: [{number, label, heading, body, items?, callout?}] }
 */
import PptxGenJSDefault from "pptxgenjs";
// CJS/ESM interop: pptxgenjs ships as CommonJS
const PptxGenJS = ((PptxGenJSDefault as unknown as { default: typeof PptxGenJSDefault }).default ?? PptxGenJSDefault) as typeof PptxGenJSDefault;
import { readFileSync, writeFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import * as bt from "./brand_tokens.js";

const __dirname = dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface Section {
  number: string;
  label: string;
  heading: string;
  body: string;
  items?: string[];
  callout?: string;
}

interface Content {
  title: string;
  subtitle?: string;
  author?: string;
  sections: Section[];
}

// ---------------------------------------------------------------------------
// Slide builders
// ---------------------------------------------------------------------------

function buildTitleSlide(pres: PptxGenJS, content: Content): void {
  const slide = pres.addSlide();

  // Dark background
  slide.background = { color: bt.hex("background_dark") };

  // Left accent bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.08, h: 7.5,
    fill: { color: bt.hex("primary") },
    line: { color: bt.hex("primary"), width: 0 },
  });

  // BBVA wordmark (text fallback when no logo asset)
  slide.addText("BBVA", {
    x: 0.4, y: 0.3, w: 2, h: 0.5,
    fontSize: 22,
    bold: true,
    color: bt.hex("primary"),
    fontFace: bt.FONTS["pptx_body"],
  });

  // Main title
  slide.addText(content.title, {
    x: 0.4, y: 2.5, w: 9, h: 1.8,
    fontSize: 44,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });

  // Subtitle
  if (content.subtitle) {
    slide.addText(content.subtitle, {
      x: 0.4, y: 4.5, w: 9, h: 0.8,
      fontSize: 20,
      color: bt.hex("secondary"),
      fontFace: bt.FONTS["pptx_body"],
      italic: true,
    });
  }

  // Author + date footer
  const footer = [content.author, new Date().getFullYear().toString()]
    .filter(Boolean)
    .join(" · ");
  slide.addText(footer, {
    x: 0.4, y: 6.8, w: 12, h: 0.4,
    fontSize: 11,
    color: bt.hex("text_light"),
    fontFace: bt.FONTS["pptx_body"],
  });
}

function buildDividerSlide(pres: PptxGenJS, section: Section): void {
  const slide = pres.addSlide();

  slide.background = { color: bt.hex("primary") };

  // Section number
  slide.addText(section.number, {
    x: 0.5, y: 1.5, w: 4, h: 2,
    fontSize: 96,
    bold: true,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_headlines"],
    alpha: 40,
  });

  // Section label
  slide.addText(section.label.toUpperCase(), {
    x: 0.5, y: 5.2, w: 10, h: 0.6,
    fontSize: 13,
    color: bt.hex("secondary"),
    fontFace: bt.FONTS["pptx_body"],
    charSpacing: 4,
  });

  // Section heading
  slide.addText(section.heading, {
    x: 0.5, y: 3.5, w: 10, h: 1.5,
    fontSize: 32,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });
}

function buildContentSlide(pres: PptxGenJS, section: Section): void {
  const slide = pres.addSlide();

  slide.background = { color: bt.hex("background") };

  // Left accent bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.08, h: 7.5,
    fill: { color: bt.hex("primary") },
    line: { color: bt.hex("primary"), width: 0 },
  });

  // Section label
  slide.addText(section.label.toUpperCase(), {
    x: 0.4, y: 0.25, w: 10, h: 0.3,
    fontSize: 10,
    color: bt.hex("primary"),
    fontFace: bt.FONTS["pptx_body"],
    charSpacing: 3,
  });

  // Heading
  slide.addText(section.heading, {
    x: 0.4, y: 0.65, w: 12, h: 1,
    fontSize: 28,
    bold: true,
    color: bt.hex("primary_dark"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });

  // Divider line
  slide.addShape(pres.ShapeType.line, {
    x: 0.4, y: 1.75, w: 12.4, h: 0,
    line: { color: bt.hex("border"), width: 1 },
  });

  // Body text
  slide.addText(section.body, {
    x: 0.4, y: 1.95, w: 7.5, h: 2.5,
    fontSize: 14,
    color: bt.hex("text_primary"),
    fontFace: bt.FONTS["pptx_body"],
    lineSpacingMultiple: 1.4,
    wrap: true,
  });

  // Bullet items
  if (section.items && section.items.length > 0) {
    const bullets = section.items.map((item) => ({
      text: item,
      options: {
        bullet: { type: "bullet" as const },
        fontSize: 13,
        color: bt.hex("text_primary"),
        fontFace: bt.FONTS["pptx_body"],
      },
    }));
    slide.addText(bullets, {
      x: 8.2, y: 1.95, w: 4.8, h: 4.5,
      wrap: true,
    });
  }

  // Callout box
  if (section.callout) {
    slide.addShape(pres.ShapeType.rect, {
      x: 0.4, y: 4.7, w: 7.5, h: 1.5,
      fill: { color: bt.hex("primary"), alpha: 8 },
      line: { color: bt.hex("primary"), width: 1 },
    });
    slide.addText(section.callout, {
      x: 0.7, y: 4.9, w: 7, h: 1.1,
      fontSize: 13,
      italic: true,
      color: bt.hex("primary_dark"),
      fontFace: bt.FONTS["pptx_body"],
      wrap: true,
    });
  }
}

function buildClosingSlide(pres: PptxGenJS, content: Content): void {
  const slide = pres.addSlide();

  slide.background = { color: bt.hex("background_dark") };

  // Accent bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.08, h: 7.5,
    fill: { color: bt.hex("accent") },
    line: { color: bt.hex("accent"), width: 0 },
  });

  slide.addText("Gracias", {
    x: 0.4, y: 2.5, w: 8, h: 1.5,
    fontSize: 52,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
  });

  slide.addText(content.title, {
    x: 0.4, y: 4.2, w: 10, h: 0.6,
    fontSize: 16,
    color: bt.hex("secondary"),
    fontFace: bt.FONTS["pptx_body"],
  });

  slide.addText("BBVA", {
    x: 0.4, y: 6.5, w: 3, h: 0.5,
    fontSize: 18,
    bold: true,
    color: bt.hex("primary"),
    fontFace: bt.FONTS["pptx_body"],
  });
}

// ---------------------------------------------------------------------------
// Main generator
// ---------------------------------------------------------------------------

export function generatePptx(content: Content, outputPath: string): void {
  const pres = new PptxGenJS();

  pres.layout = "LAYOUT_WIDE";
  pres.title  = content.title;
  pres.author = content.author ?? "BBVA";

  buildTitleSlide(pres, content);

  for (const section of content.sections) {
    buildDividerSlide(pres, section);
    buildContentSlide(pres, section);
  }

  buildClosingSlide(pres, content);

  pres.writeFile({ fileName: outputPath });
  console.log(`PPTX written: ${outputPath} (${content.sections.length} sections)`);
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

function parseCli(): { content: Content; output: string } {
  const args = process.argv.slice(2);
  const getArg = (flag: string): string | undefined => {
    const idx = args.indexOf(flag);
    return idx !== -1 ? args[idx + 1] : undefined;
  };

  const contentFile = getArg("--content");
  const output = getArg("--output") ?? "output.pptx";
  const titleArg = getArg("--title");

  if (contentFile) {
    const raw = JSON.parse(readFileSync(contentFile, "utf-8")) as Content;
    return { content: raw, output };
  }

  if (titleArg) {
    return {
      content: { title: titleArg, sections: [] },
      output,
    };
  }

  throw new Error(
    "Usage: npx tsx generate_pptx.ts --content content.json --output output.pptx\n" +
    "       npx tsx generate_pptx.ts --title 'My Deck' --output output.pptx"
  );
}

if (process.argv[1] && fileURLToPath(import.meta.url).endsWith(process.argv[1].split("/").pop()!)) {
  const { content, output } = parseCli();
  generatePptx(content, output);
}

File: src/templates/skills/bbva-brand/scripts/python/generate_pptx.py
#!/usr/bin/env python3
"""
generate_pptx.py — BBVA-branded PPTX generator (python-pptx, FALLBACK runtime).

Usage:
    python3 generate_pptx.py --content content.json --output output.pptx

Install dependency: pip install python-pptx
content.json schema: {title, subtitle?, author?, sections: [{number, label, heading, body, items?, callout?}]}
"""

import sys
import json
import argparse
from pathlib import Path
from datetime import datetime

sys.path.insert(0, str(Path(__file__).parent))
import brand_tokens as bt

from pptx import Presentation
from pptx.util import Inches, Pt
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN

SLIDE_W = Inches(float(bt.LAYOUT["slide_width_in"]))
SLIDE_H = Inches(float(bt.LAYOUT["slide_height_in"]))
MARGIN = Inches(float(bt.LAYOUT["content_margin_in"]))
BAR_W = Inches(float(bt.LAYOUT["accent_bar_width_in"]))


def _rgb(key: str) -> RGBColor:
    r, g, b = bt.rgb(key)
    return RGBColor(r, g, b)


def _set_bg(slide, key: str) -> None:
    fill = slide.background.fill
    fill.solid()
    fill.fore_color.rgb = _rgb(key)


def _add_rect(slide, x, y, w, h, color_key: str) -> None:
    shape = slide.shapes.add_shape(
        1,
        x,
        y,
        w,
        h,  # MSO_SHAPE_TYPE.RECTANGLE = 1
    )
    shape.fill.solid()
    shape.fill.fore_color.rgb = _rgb(color_key)
    shape.line.fill.background()


def _add_text(
    slide,
    text: str,
    x,
    y,
    w,
    h,
    size: int,
    color_key: str,
    font_key: str = "pptx_body",
    bold=False,
    italic=False,
    align=PP_ALIGN.LEFT,
) -> None:
    txBox = slide.shapes.add_textbox(x, y, w, h)
    tf = txBox.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    p.alignment = align
    run = p.add_run()
    run.text = text
    run.font.size = Pt(size)
    run.font.color.rgb = _rgb(color_key)
    run.font.name = bt.FONTS.get(font_key, "Arial")
    run.font.bold = bold
    run.font.italic = italic


def build_title_slide(prs: Presentation, content: dict) -> None:
    slide = prs.slides.add_slide(prs.slide_layouts[6])  # blank layout
    _set_bg(slide, "background_dark")
    _add_rect(slide, 0, 0, BAR_W, SLIDE_H, "primary")

    # BBVA wordmark
    _add_text(
        slide,
        "BBVA",
        MARGIN,
        Inches(0.3),
        Inches(2),
        Inches(0.5),
        22,
        "primary",
        "pptx_body",
        bold=True,
    )

    # Title
    _add_text(
        slide,
        content["title"],
        MARGIN,
        Inches(2.5),
        Inches(9),
        Inches(1.8),
        44,
        "white",
        "pptx_headlines",
        bold=True,
    )

    # Subtitle
    if content.get("subtitle"):
        _add_text(
            slide,
            content["subtitle"],
            MARGIN,
            Inches(4.5),
            Inches(9),
            Inches(0.8),
            20,
            "secondary",
            "pptx_body",
            italic=True,
        )

    # Footer
    author = content.get("author", "")
    year = str(datetime.now().year)
    footer = f"{author} · {year}" if author else year
    _add_text(
        slide,
        footer,
        MARGIN,
        Inches(6.8),
        Inches(12),
        Inches(0.4),
        11,
        "text_light",
        "pptx_body",
    )


def build_divider_slide(prs: Presentation, section: dict) -> None:
    slide = prs.slides.add_slide(prs.slide_layouts[6])
    _set_bg(slide, "primary")

    # Large section number (decorative)
    _add_text(
        slide,
        section["number"],
        Inches(0.5),
        Inches(1.5),
        Inches(4),
        Inches(2),
        96,
        "accent",
        "pptx_headlines",
        bold=True,
    )

    # Section heading
    _add_text(
        slide,
        section["heading"],
        Inches(0.5),
        Inches(3.5),
        Inches(10),
        Inches(1.5),
        32,
        "white",
        "pptx_headlines",
        bold=True,
    )

    # Section label
    _add_text(
        slide,
        section["label"].upper(),
        Inches(0.5),
        Inches(5.2),
        Inches(10),
        Inches(0.6),
        13,
        "secondary",
        "pptx_body",
    )


def build_content_slide(prs: Presentation, section: dict) -> None:
    slide = prs.slides.add_slide(prs.slide_layouts[6])
    _set_bg(slide, "background")
    _add_rect(slide, 0, 0, BAR_W, SLIDE_H, "primary")

    # Label
    _add_text(
        slide,
        section["label"].upper(),
        MARGIN,
        Inches(0.25),
        Inches(10),
        Inches(0.3),
        10,
        "primary",
        "pptx_body",
    )

    # Heading
    _add_text(
        slide,
        section["heading"],
        MARGIN,
        Inches(0.65),
        Inches(12),
        Inches(1),
        28,
        "primary_dark",
        "pptx_headlines",
        bold=True,
    )

    # Body
    _add_text(
        slide,
        section["body"],
        MARGIN,
        Inches(1.95),
        Inches(7.5),
        Inches(2.5),
        14,
        "text_primary",
        "pptx_body",
    )

    # Bullet items
    items = section.get("items", [])
    if items:
        txBox = slide.shapes.add_textbox(
            Inches(8.2), Inches(1.95), Inches(4.8), Inches(4.5)
        )
        tf = txBox.text_frame
        tf.word_wrap = True
        for i, item in enumerate(items):
            p = tf.paragraphs[0] if i == 0 else tf.add_paragraph()
            p.text = f"• {item}"
            p.font.size = Pt(13)
            p.font.color.rgb = _rgb("text_primary")
            p.font.name = bt.FONTS.get("pptx_body", "Arial")

    # Callout
    if section.get("callout"):
        _add_rect(slide, MARGIN, Inches(4.7), Inches(7.5), Inches(1.5), "border")
        _add_text(
            slide,
            section["callout"],
            MARGIN + Inches(0.15),
            Inches(4.9),
            Inches(7.1),
            Inches(1.1),
            13,
            "primary_dark",
            "pptx_body",
            italic=True,
        )


def build_closing_slide(prs: Presentation, content: dict) -> None:
    slide = prs.slides.add_slide(prs.slide_layouts[6])
    _set_bg(slide, "background_dark")
    _add_rect(slide, 0, 0, BAR_W, SLIDE_H, "accent")

    _add_text(
        slide,
        "Gracias",
        MARGIN,
        Inches(2.5),
        Inches(8),
        Inches(1.5),
        52,
        "white",
        "pptx_headlines",
        bold=True,
    )
    _add_text(
        slide,
        content["title"],
        MARGIN,
        Inches(4.2),
        Inches(10),
        Inches(0.6),
        16,
        "secondary",
        "pptx_body",
    )
    _add_text(
        slide,
        "BBVA",
        MARGIN,
        Inches(6.5),
        Inches(3),
        Inches(0.5),
        18,
        "primary",
        "pptx_body",
        bold=True,
    )


def generate_pptx(content: dict, output_path: str) -> None:
    prs = Presentation()
    prs.slide_width = SLIDE_W
    prs.slide_height = SLIDE_H

    build_title_slide(prs, content)
    for section in content.get("sections", []):
        build_divider_slide(prs, section)
        build_content_slide(prs, section)
    build_closing_slide(prs, content)

    prs.save(output_path)
    n = len(content.get("sections", []))
    print(f"PPTX written: {output_path} ({n} sections)")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="BBVA PPTX generator (python-pptx fallback)"
    )
    parser.add_argument("--content", required=True, help="Path to content.json")
    parser.add_argument("--output", default="output.pptx", help="Output .pptx file")
    args = parser.parse_args()

    with open(args.content, "r", encoding="utf-8") as f:
        content = json.load(f)

    generate_pptx(content, args.output)


if __name__ == "__main__":
    main()
